src/qibolab/instruments/zhinst.py

```diff
@@ -0,0 +1,1050 @@
+                        results[self.sequence[f"readout{qubit.name}"][0].pulse.qubit] = options.results_type(data)
+                    else:
+                        results[self.sequence[f"readout{qubit.name}"][0].pulse.serial] = options.results_type(
+                            data=np.array(exp_res)
+                        )
+                        results[self.sequence[f"readout{qubit.name}"][0].pulse.qubit] = options.results_type(
+                            data=np.array(exp_res)
+                        )
+
+        exp_dimensions = list(np.array(exp_res).shape)
+        if dimensions != exp_dimensions:
+            print("dimensions", dimensions, "experiment", exp_dimensions)
+            warnings.warn("dimensions not properly ordered")
+
+        # FIXME: Include this on the reports
+        # html containing the pulse sequence schedule
+        # lo.show_pulse_sheet("pulses", self.exp)
+
+        # There is no reason for disconnection and it prevents reconnection
+        # for a period of time making the software loops with execute_play_sequence crash
+        # self.disconnect()
+        return results
+
+    def sequence_zh(self, sequence, qubits, sweepers):
+        """Qibo sequence to Zurich sequence"""
+
+        "Define and assign the sequence"
+        zhsequence = defaultdict(list)
+        self.sequence_qibo = sequence
+
+        "Fill the sequences with pulses according to their lines in temporal order"
+        # TODO: Check if they invert the order if this will still work
+        for pulse in sequence:
+            zhsequence[f"{pulse.type.name.lower()}{pulse.qubit}"].append(ZhPulse(pulse))
+
+        "Mess that gets the sweeper and substitutes the pulse it sweeps in the right place"
+        for sweeper in sweepers:
+            if sweeper.parameter.name in {"amplitude", "frequency", "duration", "relative_phase"}:
+                for pulse in sweeper.pulses:
+                    aux_list = zhsequence[f"{pulse.type.name.lower()}{pulse.qubit}"]
+                    if sweeper.parameter is Parameter.frequency and pulse.type is PulseType.READOUT:
+                        self.acquisition_type = lo.AcquisitionType.SPECTROSCOPY
+                    if sweeper.parameter is Parameter.amplitude and pulse.type is PulseType.READOUT:
+                        self.acquisition_type = lo.AcquisitionType.SPECTROSCOPY
+                        self.NT_sweeps.append(sweeper)
+                        self.sweepers.remove(sweeper)
+                    for element in aux_list:
+                        if pulse == element.pulse:
+                            if isinstance(aux_list[aux_list.index(element)], ZhPulse):
+                                aux_list[aux_list.index(element)] = ZhSweeper(pulse, sweeper, qubits[pulse.qubit])
+                            elif isinstance(aux_list[aux_list.index(element)], ZhSweeper):
+                                aux_list[aux_list.index(element)].add_sweeper(sweeper, qubits[pulse.qubit])
```

Comment:
If you can split this into a separate function would be great!

Reply:
I cant think of a way right now